heron/api/src/java/com/twitter/heron/api/state/State.java

```diff
@@ -0,0 +1,26 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.api.state;
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * State represents the state interface as seen by stateful bolts and spouts.
+ * In Heron, state gives a notional Key/Value interface along with the
+ * ability to iterate over the key/values
+ */
+public interface State extends Map<Serializable, Serializable> {
```

Comment:
`Map<<K extends Serializable>, <V extends Serializable>>` says that we should take any kind of state as long as it is serializable. The difference is how the users will write their code, not how the system deals with serializables. For the spout and bolt in your example, one would be able to work with `<String, Integer>` while the other would be able to work with `<String, String>`. This allows the user to write type safe code with compile time checks, while still allowing the system to persist any types of serializables.

To do this we would have something like this, where the user knows they can work with `<String, Integer>' for example:
```
interface IStatefulComponent<String, Integer> {
  initState(State<String, Integer> state);
}
```